### PR TITLE
8237913: G1CollectedHeap::heap_region_containing shouldn't be a template

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that removes the templates and casts around `G1CollectedHeap::heap_region_containing`?

Testing: tier1-5

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8237913](https://bugs.openjdk.org/browse/JDK-8237913): G1CollectedHeap::heap_region_containing shouldn't be a template

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Sangheon Kim](https://openjdk.org/census#sangheki) (@sangheon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9572/head:pull/9572` \
`$ git checkout pull/9572`

Update a local copy of the PR: \
`$ git checkout pull/9572` \
`$ git pull https://git.openjdk.org/jdk pull/9572/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9572`

View PR using the GUI difftool: \
`$ git pr show -t 9572`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9572.diff">https://git.openjdk.org/jdk/pull/9572.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1CollectedHeap.hpp [hunk 1]
```diff
@@ -1121,14 +1121,12 @@ class G1CollectedHeap : public CollectedHeap {
...
+  // Returns the HeapRegion that contains addr. addr must not be nullptr.
+  inline HeapRegion* heap_region_containing(const void* addr) const;
 
-  // Returns the HeapRegion that contains addr, or NULL if that is an uncommitted
-  // region. addr must not be NULL.
-  template <class T>
-  inline HeapRegion* heap_region_containing_or_null(const T addr) const;
+  // Returns the HeapRegion that contains addr, or nullptr if that is an uncommitted
+  // region. addr must not be nullptr.
+  inline HeapRegion* heap_region_containing_or_null(HeapWord* addr) const;
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1CollectedHeap.hpp:1129]: Why does `heap_region_containing` take `void*` but `heap_region_containing_or_null` takes `HeapWord*`.  Seems like they should be the same (probably `void*` for both?).  And why did the `const` qualifier get dropped for the or_null variant?

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CollectedHeap.hpp:1129]: `heap_region_containing_or_null` only needs the `HeapWord*` variant - I thought to specialize them as much as possible; dropping of the const has been unintenional. I will fix and regularize these two methods. Thanks.

**Review by kimbarrett — APPROVED** [review]
Looks good.

File: src/hotspot/share/gc/g1/g1CollectedHeap.inline.hpp [hunk 2]
```diff
@@ -97,21 +97,15 @@ inline HeapWord* G1CollectedHeap::bottom_addr_for_region(uint index) const {
...
+  return _hrm.addr_to_region((HeapWord*)addr);
 }
 
-template <class T>
-inline HeapRegion* G1CollectedHeap::heap_region_containing_or_null(const T addr) const {
-  assert(addr != NULL, "invariant");
-  assert(is_in_reserved((const void*) addr),
-         "Address " PTR_FORMAT " is outside of the heap ranging from [" PTR_FORMAT " to " PTR_FORMAT ")",
-         p2i((void*)addr), p2i(reserved().start()), p2i(reserved().end()));
+inline HeapRegion* G1CollectedHeap::heap_region_containing_or_null(HeapWord* addr) const {
```
[the remark this answers is not in the document]
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CollectedHeap.inline.hpp]: The `addr_to_region` call already checks whether the given address is in the heap and `nullptr` is always outside the heap, so redundant anyway.

**Review by sangheon — APPROVED** [review]
Looks good.

**Comment by tschatzl** [comment]
Thanks @sangheon @kimbarrett for your reviews

/integrate

**Pull request closed without merging**